.github/workflows/main.yml

```diff
@@ -0,0 +1,109 @@
+name: CI Workflow
+
+on:
+  push:
+    branches: 
+      - '*'
+  pull_request:
+    branches:
+      - 'main'
+      - 'dev'
+
+jobs:
+  unit_tests:
+    name: unit_tests
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Python 
+        uses: actions/setup-python@v3
+        with:
+          python-version: '3.10'
+  
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+    
+      - name: Run all unit tests
+        run: |
+            pytest test/unit --doctest-modules --junitxml=junit/test-results.xml
+
+  # Build Images (not pubslish) on pull to dev & main
+  docker_build:
+    needs: unit_tests
+    name: docker_build
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request' && (github.head_ref == 'dev' || github.head_ref == 'main')
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Build Docker Images 
+        run: |
+          docker build --file ./microservices/ingestservice/docker/Dockerfile ./microservices/ingestservice
+          echo "ingest service image build OK"
+          docker build --file ./microservices/searchservice/docker/Dockerfile ./microservices/searchservice
+          echo "search service image build OK"
+          docker build --file ./microservices/embeddingservice/docker/Dockerfile ./microservices/embeddingservice
+          echo "embedding service image build OK"
+
+  
+  docker_build_publish_dev:
```

Comment:
Might be out of scope, but I think it would be a good idea to have seperate pipelines for every microservice in order to be able to deploy them seperately. We might want to run unit tests for every service in the future. In this combined setup we will be running tests unneccessarily.

Reply:
Good point - My plan was to break it up as soon as we add the deployment logic. The setup for now only builds and publishes. So I would leave it as is for now.